plugins/backstage-plugin-coder/src/api/CoderClient.ts

```diff
@@ -0,0 +1,348 @@
+  CoderSdk & {
+    getWorkspacesByRepo: (
+      request: WorkspacesRequest,
+      config: CoderWorkspacesConfig,
+    ) => Promise<WorkspacesResponse>;
+  }
+>;
+
+type CoderClientApi = Readonly<{
+  sdk: BackstageCoderSdk;
+
+  /**
+   * Validates a new token, and loads it only if it is valid.
+   * Return value indicates whether the token is valid.
+   */
+  syncToken: (newToken: string) => Promise<boolean>;
+
+  /**
+   * Cleans up a client instance, removing its links to all external systems.
+   */
+  cleanupClient: () => void;
+}>;
+
+const sharedCleanupAbortReason = new DOMException(
+  'Coder Client instance has been manually cleaned up',
+  'AbortError',
+);
+
+// Can't make this value readonly at the type level because it has
+// non-enumerable properties, and Object.freeze causes errors. Just have to
+// treat this like a constant
+export const disabledClientError = new Error(
+  'Requests have been disabled for this client. Please create a new client',
+);
+
+type ConstructorInputs = Readonly<{
+  initialToken?: string;
+  requestTimeoutMs?: number;
+
+  apis: Readonly<{
+    urlSync: UrlSync;
+    identityApi: IdentityApi;
+  }>;
+}>;
+
+export class CoderClient implements CoderClientApi {
+  private readonly urlSync: UrlSync;
+  private readonly identityApi: IdentityApi;
+  private readonly axios: AxiosInstance;
+
+  private readonly requestTimeoutMs: number;
+  private readonly cleanupController: AbortController;
+  private readonly trackedEjectionIds: Set<number>;
+
+  private loadedSessionToken: string | undefined;
+  readonly sdk: BackstageCoderSdk;
+
+  constructor(inputs: ConstructorInputs) {
+    const {
+      apis,
+      initialToken,
+      requestTimeoutMs = DEFAULT_REQUEST_TIMEOUT_MS,
+    } = inputs;
+    const { urlSync, identityApi } = apis;
+
+    this.urlSync = urlSync;
+    this.identityApi = identityApi;
+    this.axios = globalAxios.create();
+
+    this.loadedSessionToken = initialToken;
+    this.requestTimeoutMs = requestTimeoutMs;
+
+    this.cleanupController = new AbortController();
+    this.trackedEjectionIds = new Set();
+
+    this.sdk = this.getBackstageCoderSdk(this.axios);
+    this.addBaseRequestInterceptors();
+  }
+
+  private addRequestInterceptor(
+    requestInterceptor: (
+      config: RequestConfig,
+    ) => RequestConfig | Promise<RequestConfig>,
+    errorInterceptor?: (error: unknown) => unknown,
+  ): number {
+    const ejectionId = this.axios.interceptors.request.use(
+      requestInterceptor,
+      errorInterceptor,
+    );
+
+    this.trackedEjectionIds.add(ejectionId);
+    return ejectionId;
+  }
+
+  private removeRequestInterceptorById(ejectionId: number): boolean {
+    const sizeBeforeRemoval = this.trackedEjectionIds.size;
+
+    this.axios.interceptors.request.eject(ejectionId);
+    if (this.trackedEjectionIds.has(ejectionId)) {
+      this.trackedEjectionIds.delete(ejectionId);
+    }
+
+    return sizeBeforeRemoval !== this.trackedEjectionIds.size;
```

Comment:
Yeah, good call on the size check – your suggestion's a lot more straightforward

As for whether the method makes sense, I guess my gut feeling is that if I'm going to have a special add method for the interceptors, it makes sense to have an obvious way to remove them, too. Plus, since it's only a private method, it'd be easy to remove down the line, too